### DISCUSSION TITLE
PHP Architecture - namespaces

### DISCUSSION THREAD

**Discussion by Sephsekla** [WordPress/wp-feature-notifications#115] [PHP API]
Looking at the current PHP skeleton, we are using prefixes rather than namespaces for class and function names. 

While there may be some value to having some functions in the global namespace, in line with the project goal to take a more modern approach to development I would like to update this to use namespaces, along with a more robust approach to inline documentation.

E.g.

```php
<?php

interface WP_Notify_Notification_Controller {

}

```

becomes something closer to 

```php
<?php
/**
 * REST API controller to get and set notifications.
 *
 * @package wordpress/wp-feature-notifications
 */

namespace WP\Notifications\REST;

/**
 * Some useful docblock
 */
interface Notification_Controller {

}

```

**Comment by JasonTheAdams** [comment]
Definitely agree in the use of namespaces and hope it paves the way for far better auto-loading in the WP future. Since this is aimed at core, do you know what the current state of namespaces is in core? Would this potentially be a problem later on when this is pitched for core?

**Reply by JasonTheAdams** [reply]
Following up on the "it’s a good idea, as long as we can put it into core" conclusion made in Slack, the last detail I'll add is to probably avoid (as much as it kills me) autoloading since it's not a concept in core — and I doubt this feature will be the one to introduce it. Each class will need to be manually included. I mean, if someone wants to reach out to a committer and see if we can get their blessing to attempt introducing auto-loading, then by all means.

**Reply by Sephsekla** [reply]
I don't think a lack of autoloading is the end of the world, although I imagine the feasibility of it would depend on whether we're using Composer or writing something to just load files matching a pattern.

I'd question whether everything needs to be a class tbh - OOP makes sense for a large amount of it, but there will also be a fair amount of functionality which might be better served as namespaced functions. The WP ecosystem tends to have a habit of using classes as pseudo-namespaces, which isn't needed if we're actually using namespaces.

**Reply by JasonTheAdams** [reply]
From Slack it sounds like autoloading is being explored elsewhere in WP, so I'd say it's worth proceeding. Apparently the [Request](https://github.com/WordPress/Requests) package even uses PSR-4, which is pretty awesome. That would be ideal!

As far as the balance between functions and classes, that really gets into architectural philosophy, wherein there isn't a "right" answer. At GiveWP we follow a pretty strict OOP server-side philosophy, and use almost exclusively functions in JS. We use the Action pattern — invokable classes with DI support — in place of most functions, reserving functions to be exclusively pure and more like helpers.

I do agree that most classes in WP aren't really OOP, they're basically just static facades — or a self-contained singleton class, which is basically a static facade. The "WP way" of doing PHP is still pretty tied to what was possible at PHP 5.2 and that mindset. In my opinion the effort should to be improve at OOP, but understandably that's a greater endeavor as it requires improving OOP competency overall, which has practical challenges. For the sake of this feature, I would hope to have the internals follow SOLID OOP principles, while making sure we provide functions and simpler such so the APIs themselves are more approachable.

**Reply by johnhooks** [reply]
I just found this thread! I've converted everything over to namespaces. I will look at the Request package @JasonTheAdams mentioned. If/when we were able to convert to PSR-4 autoloading, the changes to the repo and lint rules are mostly around the file structure not the code itself.

**Reply by johnhooks** [reply]
I'll also research the object orientated patterns. One issue I'm having, I need to dive into core to see how things work, but I'm picking up habits from what I see. Please point me in the right direction where ever you see room for improvement.
